sdk/java/pulumi/src/main/java/com/pulumi/resources/internal/Stack.java

```diff
@@ -0,0 +1,109 @@
+     */
+    @InternalUse
+    public static final String RootPulumiStackTypeName = "pulumi:pulumi:Stack";
+
+    /**
+     * The outputs of this stack, if the <code>callback</code> callback exited normally.
+     */
+    private final Output<Map<String, Output<?>>> outputs;
+
+    /**
+     * Create a Stack with stack outputs defined in derived class constructor.
+     *
+     * @param projectName             the current project name
+     * @param stackName               the current stack name
+     * @param resourceTransformations optional list of transformations to apply to this stack's resources during construction.
+     *                                The transformations are applied in order, and are applied after all the transformations of custom
+     *                                and component resources in the stack.
+     * @param outputs                 the stack outputs
+     */
+    private Stack(
+            String projectName,
+            String stackName,
+            List<ResourceTransformation> resourceTransformations,
+            Output<Map<String, Output<?>>> outputs
+    ) {
+        super(
+                RootPulumiStackTypeName,
+                rootPulumiStackName(projectName, stackName),
+                ComponentResourceOptions.builder().resourceTransformations(
+                        resourceTransformations
+                ).build()
+        );
+        this.outputs = requireNonNull(outputs);
+    }
+
+    private static String rootPulumiStackName(String projectName, String stackName) {
+        return String.format("%s-%s", requireNonNull(projectName), requireNonNull(stackName));
+    }
+
+    @InternalUse
+    public Output<Map<String, Output<?>>> getOutputs() {
+        return this.outputs;
+    }
+
+    @InternalUse
+    public static Function<Supplier<Map<String, Output<?>>>, Stack> factory(
+            String projectName,
+            String stackName,
+            List<ResourceTransformation> resourceTransformations
+    ) {
+        // Stack initialization requires specific sequence of events:
+        // - create a Stack instance (note that the super constructor calls readOrRegisterResource)
+        // - set the Stack instance in the global state (user code will try to reach it via Resource constructor)
+        // - register the Stack outputs in the Pulumi engine
+        // - call the user code callback (asynchronously to avoid any additional cyclic dependency problems)
+        // - complete the Stack initialization (by completing the lazyFuture)
+        return (Supplier<Map<String, Output<?>>> callback) -> {
+            // incomplete future that will be completed after the Stack is constructed
+            var lazyFuture = new CompletableFuture<Map<String, Output<?>>>();
+            // incomplete outputs that will be completed when lazyFuture completes
+            var lazyOutputs = Output.of(lazyFuture);
+            // Stack must be the first Resource to be constructed,
```

Comment:
Not a fan of commenting every line, that's diminishing returns for comments the code speaks for itself ideally at this level.

Reply:
I agree, ideally code should speak for itself. After struggling to understand the sequence and the reasons behind it myself I wanted to help everyone else who might need to do the same in the future.
Would rewriting the comments into a one big description on to be more readable for you?

Reply:
I've moved the comments/clarify, hope it helps.